### ISSUE TITLE
tests/drivers/uart/uart_async_api fails on nrf52840dk_nrf52840 (and additional platforms)

### ISSUE THREAD

**Issue by pabigot** [zephyrproject-rtos/zephyr#29092]
The test on this platform produces:
```
*** Booting Zephyr OS build zephyr-v2.4.0-536-g9072f78abf15  ***
Running test suite uart_async_test
===================================================================
START - test_single_read_setup
 PASS - test_single_read_setup
===================================================================
START - test_single_read
E: ***** MPU FAULT *****
E:   Data Access Violation
E:   MMFAR Address: 0x20001490
E: r0/a1:  0x00000001  r1/a2:  0x0001089e  r2/a3:  0x00010856
E: r3/a4:  0x00000049 r12/ip:  0x0000e7ed r14/lr:  0x00000953
E:  xpsr:  0x21000000
E: Faulting instruction address (r15/pc): 0x00000956
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Current thread: 0x20001000 (unknown)
E: Halting system
```

Though it's not documented I believe I'm supposed to short P1.01-P1.02 and P0.5-P0.7 as a loopback configuration, but the behavior is unchanged whether or not this is done.

**Comment by pabigot** [comment]
The test fails with v2.4.0.  @carlescufi 

The tests passes with v2.3.0.

With bisect it runs but fails several tests at 63c3e153d65a7ae2ece0587dc7e4a4be93553cb9

Calling that good, the MPU fault was introduced by 26cbd6a58c8346e21cee6efa3c0f7a29457afc92

That's fixed in #29093, but then the test fails when several checks for ready events time out.  I'll continue to look into that.

**Comment by pabigot** [comment]
The remaining failures are shown here:
<details><summary>Test failures for ready events</summary>

```
===================================================================
START - test_chained_write

    Assertion failed at ../src/test_uart_async.c:492: test_chained_write: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_chained_write
===================================================================
START - test_long_buffers_setup
 PASS - test_long_buffers_setup
===================================================================
START - test_long_buffers

    Assertion failed at ../src/test_uart_async.c:564: test_long_buffers: (k_sem_take(&rx_rdy, K_MSEC(200)) not equal to 0)
RX_RDY timeout
 FAIL - test_long_buffers
===================================================================
START - test_write_abort_setup
 PASS - test_write_abort_setup
===================================================================
START - test_write_abort

    Assertion failed at ../src/test_uart_async.c:350: test_write_abort: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_write_abort
===================================================================
START - test_forever_timeout_setup
 PASS - test_forever_timeout_setup
===================================================================
START - test_forever_timeout

    Assertion failed at ../src/test_uart_async.c:426: test_forever_timeout: (k_sem_take(&rx_rdy, K_MSEC(100)) not equal to 0)
RX_RDY timeout
 FAIL - test_forever_timeout
===================================================================
Test suite uart_async_test failed.
===================================================================
PROJECT EXECUTION FAILED
```
</details>

Serendipitously confirming my belief that the use of `k_delayed_work` there was wrong, these are fixed by #28891 which changes the driver to use to the new delayable work API.
